File: packages/runpane-py/src/runpane/peers.py
from __future__ import annotations

import json
import os
import sys
from typing import TYPE_CHECKING

from .daemon_client import invoke_daemon

if TYPE_CHECKING:
    from .cli import ParsedArgs


def run_peers(parsed: ParsedArgs) -> int:
    action = parsed.command.split(" ")[1]
    if parsed.dry_run:
        raise ValueError("Peer commands do not support --dry-run; use self, list or an unclaimed inbox to inspect.")
    if parsed.follow and action != "wait":
        raise ValueError("--follow is supported only by peers wait.")
    if parsed.follow and parsed.timeout_ms == 0:
        raise ValueError("--follow requires a positive timeout.")
    if parsed.panel_input is not None and parsed.panel_input_file is not None:
        raise ValueError("Use either --text or --input-file.")
    text = parsed.panel_input
    if parsed.panel_input_file == "-":
        text = sys.stdin.read()
    elif parsed.panel_input_file is not None:
        with open(parsed.panel_input_file, encoding="utf-8") as source:
            text = source.read()
    request = {key: value for key, value in {
        "action": action,
        "peer": parsed.peer or os.environ.get("PANE_PEER_ID") or os.environ.get("PANE_PANEL_ID"),
        "to": parsed.peer_to, "id": parsed.message_id, "agent": parsed.agent_label,
        "receiver": parsed.receiver, "status": parsed.reply_status, "text": text,
        "claim": parsed.claim, "includeReceived": parsed.include_received,
        "after": parsed.after_revision, "timeoutMs": parsed.timeout_ms,
        "limit": parsed.limit, "confirmed": parsed.yes,
    }.items() if value is not None}
    while True:
        result = invoke_daemon("runpane:peers", [request], pane_dir=parsed.pane_dir,
                               timeout_ms=(parsed.timeout_ms if parsed.timeout_ms is not None else 60_000) + 10_000,
                               event_include=[])
        if not isinstance(result, dict) or result.get("protocolVersion") != 1:
            raise ValueError("Unsupported peer protocol response.")
        if parsed.follow and result.get("timedOut") is True:
            continue
        print(json.dumps(result, indent=None if parsed.json else 2, ensure_ascii=False), flush=True)
        return 0

File: packages/runpane/src/peers.ts
import fs from 'node:fs';
import { invokeDaemon } from './daemonClient';
import { boundary } from './boundaryDecoder';
import type { ParsedArgs } from './commands';

export async function runPeers(parsed: ParsedArgs): Promise<number> {
  const action = parsed.command.split(' ')[1];
  if (parsed.dryRun) throw new Error('Peer commands do not support --dry-run; use self, list or an unclaimed inbox to inspect.');
  if (parsed.follow && action !== 'wait') throw new Error('--follow is supported only by peers wait.');
  if (parsed.follow && parsed.timeoutMs === 0) throw new Error('--follow requires a positive timeout.');
  if (parsed.panelInput !== undefined && parsed.panelInputFile !== undefined) throw new Error('Use either --text or --input-file.');
  const text = parsed.panelInputFile !== undefined
    ? fs.readFileSync(parsed.panelInputFile === '-' ? 0 : parsed.panelInputFile, 'utf8')
    : parsed.panelInput;
  const request = {
    action, peer: parsed.peer ?? process.env.PANE_PEER_ID ?? process.env.PANE_PANEL_ID,
    to: parsed.peerTo, id: parsed.messageId, agent: parsed.agentLabel, receiver: parsed.receiver,
    status: parsed.replyStatus, text, claim: parsed.claim, includeReceived: parsed.includeReceived,
    after: parsed.afterRevision, timeoutMs: parsed.timeoutMs, limit: parsed.limit, confirmed: parsed.yes,
  };
  do {
    const result = await invokeDaemon('runpane:peers', [request], boundary.jsonObject, {
      paneDir: parsed.paneDir, timeoutMs: (parsed.timeoutMs ?? 60_000) + 10_000, eventInclude: [],
    });
    if (result.protocolVersion !== 1) throw new Error('Unsupported peer protocol response.');
    if (parsed.follow && result.timedOut === true) continue;
    // Compact JSON is the common interface for humans, agents and native bridges.
    console.log(JSON.stringify(result, null, parsed.json ? undefined : 2));
    return 0;
  } while (parsed.follow);
  return 0;
}
